src/Event/Preprocess/ContentEntityPreprocessEvent.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace Drupal\hook_event_dispatcher\Event\Preprocess;
+
+/**
+ * Class EntityPreprocessEvent.
+ */
+class ContentEntityPreprocessEvent extends AbstractPreprocessEvent implements ComposedPreprocessEventInterface {
+
+  /**
+   * Get the hook name.
+   *
+   * @return string
+   *   Hook name.
+   */
+  public static function getHook() {
+    return 'entity';
+  }
+
+  /**
+   * Get the hook name.
+   *
+   * @return string
+   *   Hook name.
+   */
+  public function getComposedName() {
+    /** @var \Drupal\hook_event_dispatcher\Event\Preprocess\Variables\ContentEntityEventVariables $vars */
+    $vars = $this->getVariables();
+    return self::name() . '.' . $vars->getEntityType();
```

Comment:
Won't this make the event `preprocess_node.node`?

Reply:
This method won't be called for Nodes, Paragraphs, ... due to a check in the service:

`$factory = $factory = $this->mapper->getFactory($hook) ? NULL : $this->mapper->getFactory('entity');`

But thinking about this, it would make more sense to override the getComposedName() on the already existing Content Entity Events and just return `self::name()` there so the check can be removed in the service.

Reply:
Looked at it some more the reason why I added the check there is that `createAndDispatchKnownEvent()` has already called the node factory and else it would fire 2 events for the node (preprocess_node , preprocess_entity.node).  

But I think I can solve it this way : 

Only fire general hook when it's not an entity.
```
/**
   * Create and dispatch the event.
   *
   * @param string $hook
   *   The hook name.
   * @param array $variables
   *   Variables.
   */
  public function createAndDispatchKnownEvents($hook, array &$variables) {
    if ($this->isContentEntityHook($variables)) {
      $this->createAndDispatchEntityEvent($hook, $variables);
      $this->createAndDispatchEntityBundleEvent($variables);
    } 
    else {
      $this->createAndDispatchKnownEvent($hook, $variables);
    }
  }
```

Change the mapper to look first for a more detailed factory and ales use the general
```
$factory = $this->mapper->getFactory($hook) ? $this->mapper->getFactory($hook) : $this->mapper->getFactory('entity');
```

Override the `getComposedName();` in all known entities to just return the `self::name();`